src/coreclr/vm/riscv64/stubs.cpp

```diff
@@ -1525,68 +1534,461 @@ static void LoadRegPair(BYTE* p, int reg1, int reg2, UINT32 offset)
+    STANDARD_VM_CONTRACT;
+
+    *(DWORD*)p = 0x0000297;// auipc t0, 0
+    p += 4;
+    *(DWORD*)p = 0x0102b503;// ld a0, 16(t0)
+    p += 4;
+    *(DWORD*)p = 0x0182b283;// ld t0, 24(t0)
+    p += 4;
+    *(DWORD*)p = 0x00028067;// jr t0
+    p += 4;
+
+    // label:
+    // arg
+    *(TADDR*)p = arg;
+    p += 8;
+    // target
+    *(PCODE*)p = target;
+    p += 8;
 }
 
 PCODE DynamicHelpers::CreateHelperWithArg(LoaderAllocator * pAllocator, TADDR arg, PCODE target)
 {
-    _ASSERTE(!"RISCV64: not implementation on riscv64!!!");
-    return NULL;
+    STANDARD_VM_CONTRACT;
+
+    BEGIN_DYNAMIC_HELPER_EMIT(32);
+
+    EmitHelperWithArg(p, rxOffset, pAllocator, arg, target);
+
+    END_DYNAMIC_HELPER_EMIT();
 }
 
 PCODE DynamicHelpers::CreateHelper(LoaderAllocator * pAllocator, TADDR arg, TADDR arg2, PCODE target)
 {
-    _ASSERTE(!"RISCV64: not implementation on riscv64!!!");
-    return NULL;
+    STANDARD_VM_CONTRACT;
+
+    BEGIN_DYNAMIC_HELPER_EMIT(48);
+
+    *(DWORD*)p = 0x00000297;// auipc t0, 0
+    p += 4;
+   *(DWORD*)p = 0x0182b503;// ld a0, 24(t0)
+    p += 4;
+    *(DWORD*)p = 0x0202b583;// ld a1, 32(t0)
+    p += 4;
+    *(DWORD*)p = 0x0282b283;// ld t0, 40(t0)
+    p += 4;
+    *(DWORD*)p = 0x00028067;// jr t0
+    p += 4;
+
+    // nop, padding to make 8 byte aligned
+    *(DWORD*)p = 0x00000013;
+    p += 4;
+
+    // label:
+    // arg
+    *(TADDR*)p = arg;
+    p += 8;
+    // arg2
+    *(TADDR*)p = arg2;
+    p += 8;
+    // target
+    *(TADDR*)p = target;
+    p += 8;
+
+    END_DYNAMIC_HELPER_EMIT();
 }
 
 PCODE DynamicHelpers::CreateHelperArgMove(LoaderAllocator * pAllocator, TADDR arg, PCODE target)
 {
-    _ASSERTE(!"RISCV64: not implementation on riscv64!!!");
-    return NULL;
+    STANDARD_VM_CONTRACT;
+
+    BEGIN_DYNAMIC_HELPER_EMIT(40);
+
+    *(DWORD*)p = 0x00000297;// auipc t0, 0
+    p += 4;
+    *(DWORD*)p = 0x00050593;// mv a1, a0
+    p += 4;
+    *(DWORD*)p = 0x0182b503;// ld a0, 24(t0)
+    p += 4;
+    *(DWORD*)p = 0x0202b283;// ld t0, 32(t0)
+    p += 4;
+    *(DWORD*)p = 0x00028067;// jr t0
+    p += 4;
+
+    // nop, padding to make 8 byte aligned
+    *(DWORD*)p = 0x00000013;
+    p += 4;
+
+    // label:
+    // arg
+    *(TADDR*)p = arg;
+    p += 8;
+    // target
+    *(TADDR*)p = target;
+    p += 8;
+
+    END_DYNAMIC_HELPER_EMIT();
 }
 
 PCODE DynamicHelpers::CreateReturn(LoaderAllocator * pAllocator)
 {
-    _ASSERTE(!"RISCV64: not implementation on riscv64!!!");
-    return NULL;
+    STANDARD_VM_CONTRACT;
+
+    BEGIN_DYNAMIC_HELPER_EMIT(4);
+
+    *(DWORD*)p = 0x00008067;// ret
+    p += 4;
+
+    END_DYNAMIC_HELPER_EMIT();
 }
 
 PCODE DynamicHelpers::CreateReturnConst(LoaderAllocator * pAllocator, TADDR arg)
 {
-    _ASSERTE(!"RISCV64: not implementation on riscv64!!!");
-    return NULL;
+    STANDARD_VM_CONTRACT;
+
+    BEGIN_DYNAMIC_HELPER_EMIT(24);
+
+    *(DWORD*)p = 0x00000297;// auipc t0, 0
+    p += 4;
+    *(DWORD*)p = 0x0102b503;// ld a0, 16(t0)
+    p += 4;
```

Comment:
I think this could be optimized to ld a0, 8(pc); the offset changed 16->8 because if you remove the auipc, the nop below is also unnecessary.